Support compilation against OpenCV 4

I tried to compile icub-main with OpenCV 4, and this is the error message with icub-main: 
~~~
2020-01-13T16:59:39.2925890Z Scanning dependencies of target stereoCalib
2020-01-13T16:59:39.3277190Z [ 74%] Building CXX object src/tools/stereoCalib/CMakeFiles/stereoCalib.dir/src/main.cpp.o
2020-01-13T16:59:41.9945860Z [ 74%] Building CXX object src/tools/stereoCalib/CMakeFiles/stereoCalib.dir/src/stereoCalibModule.cpp.o
2020-01-13T16:59:44.6748180Z [ 74%] Building CXX object src/tools/stereoCalib/CMakeFiles/stereoCalib.dir/src/stereoCalibThread.cpp.o
2020-01-13T16:59:46.4877660Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:190:80: error: use of undeclared identifier 'CV_CALIB_CB_ADAPTIVE_THRESH'; did you mean 'CALIB_CB_ADAPTIVE_THRESH'?
2020-01-13T16:59:46.4877920Z                     foundL = findChessboardCorners(Left, boardSize, pointbufL, CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.4878060Z                                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.4878170Z                                                                                CALIB_CB_ADAPTIVE_THRESH
2020-01-13T16:59:46.4881460Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:255:8: note: 'CALIB_CB_ADAPTIVE_THRESH' declared here
2020-01-13T16:59:46.4881600Z enum { CALIB_CB_ADAPTIVE_THRESH = 1,
2020-01-13T16:59:46.4881690Z        ^
2020-01-13T16:59:46.4882280Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:190:110: error: use of undeclared identifier 'CV_CALIB_CB_FAST_CHECK'; did you mean 'CALIB_CB_FAST_CHECK'?
2020-01-13T16:59:46.4882450Z                     foundL = findChessboardCorners(Left, boardSize, pointbufL, CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.4882580Z                                                                                                              ^~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.4883000Z                                                                                                              CALIB_CB_FAST_CHECK
2020-01-13T16:59:46.4883540Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:258:8: note: 'CALIB_CB_FAST_CHECK' declared here
2020-01-13T16:59:46.4883650Z        CALIB_CB_FAST_CHECK      = 8,
2020-01-13T16:59:46.4883730Z        ^
2020-01-13T16:59:46.4884350Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:190:135: error: use of undeclared identifier 'CV_CALIB_CB_NORMALIZE_IMAGE'; did you mean 'CALIB_CB_NORMALIZE_IMAGE'?
2020-01-13T16:59:46.4884480Z                     foundL = findChessboardCorners(Left, boardSize, pointbufL, CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.4884610Z                                                                                                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.4884710Z                                                                                                                                       CALIB_CB_NORMALIZE_IMAGE
2020-01-13T16:59:46.4885200Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:256:8: note: 'CALIB_CB_NORMALIZE_IMAGE' declared here
2020-01-13T16:59:46.4885300Z        CALIB_CB_NORMALIZE_IMAGE = 2,
2020-01-13T16:59:46.4885380Z        ^
2020-01-13T16:59:46.5302790Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:191:81: error: use of undeclared identifier 'CV_CALIB_CB_ADAPTIVE_THRESH'; did you mean 'CALIB_CB_ADAPTIVE_THRESH'?
2020-01-13T16:59:46.5303450Z                     foundR = findChessboardCorners(Right, boardSize, pointbufR, CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.5303580Z                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.5303670Z                                                                                 CALIB_CB_ADAPTIVE_THRESH
2020-01-13T16:59:46.5304200Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:255:8: note: 'CALIB_CB_ADAPTIVE_THRESH' declared here
2020-01-13T16:59:46.5304310Z enum { CALIB_CB_ADAPTIVE_THRESH = 1,
2020-01-13T16:59:46.5304390Z        ^
2020-01-13T16:59:46.5304930Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:191:111: error: use of undeclared identifier 'CV_CALIB_CB_FAST_CHECK'; did you mean 'CALIB_CB_FAST_CHECK'?
2020-01-13T16:59:46.5305070Z                     foundR = findChessboardCorners(Right, boardSize, pointbufR, CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.5305320Z                                                                                                               ^~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.5305420Z                                                                                                               CALIB_CB_FAST_CHECK
2020-01-13T16:59:46.5305900Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:258:8: note: 'CALIB_CB_FAST_CHECK' declared here
2020-01-13T16:59:46.5306000Z        CALIB_CB_FAST_CHECK      = 8,
2020-01-13T16:59:46.5306060Z        ^
2020-01-13T16:59:46.5306610Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:191:136: error: use of undeclared identifier 'CV_CALIB_CB_NORMALIZE_IMAGE'; did you mean 'CALIB_CB_NORMALIZE_IMAGE'?
2020-01-13T16:59:46.5306750Z                     foundR = findChessboardCorners(Right, boardSize, pointbufR, CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.5307040Z                                                                                                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.5307170Z                                                                                                                                        CALIB_CB_NORMALIZE_IMAGE
2020-01-13T16:59:46.5307710Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:256:8: note: 'CALIB_CB_NORMALIZE_IMAGE' declared here
2020-01-13T16:59:46.5307800Z        CALIB_CB_NORMALIZE_IMAGE = 2,
2020-01-13T16:59:46.5307870Z        ^
2020-01-13T16:59:46.5778150Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:312:80: error: use of undeclared identifier 'CV_CALIB_CB_ADAPTIVE_THRESH'; did you mean 'CALIB_CB_ADAPTIVE_THRESH'?
2020-01-13T16:59:46.5778450Z                     foundL = findChessboardCorners(Left, boardSize, pointbufL, CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.5778630Z                                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.5778730Z                                                                                CALIB_CB_ADAPTIVE_THRESH
2020-01-13T16:59:46.5779300Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:255:8: note: 'CALIB_CB_ADAPTIVE_THRESH' declared here
2020-01-13T16:59:46.5779630Z enum { CALIB_CB_ADAPTIVE_THRESH = 1,
2020-01-13T16:59:46.5789830Z        ^
2020-01-13T16:59:46.5790750Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:312:110: error: use of undeclared identifier 'CV_CALIB_CB_FAST_CHECK'; did you mean 'CALIB_CB_FAST_CHECK'?
2020-01-13T16:59:46.5791360Z                     foundL = findChessboardCorners(Left, boardSize, pointbufL, CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.5791470Z                                                                                                              ^~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.5791580Z                                                                                                              CALIB_CB_FAST_CHECK
2020-01-13T16:59:46.5792140Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:258:8: note: 'CALIB_CB_FAST_CHECK' declared here
2020-01-13T16:59:46.5792240Z        CALIB_CB_FAST_CHECK      = 8,
2020-01-13T16:59:46.5792310Z        ^
2020-01-13T16:59:46.5792890Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:312:135: error: use of undeclared identifier 'CV_CALIB_CB_NORMALIZE_IMAGE'; did you mean 'CALIB_CB_NORMALIZE_IMAGE'?
2020-01-13T16:59:46.5793040Z                     foundL = findChessboardCorners(Left, boardSize, pointbufL, CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.5793160Z                                                                                                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.5793270Z                                                                                                                                       CALIB_CB_NORMALIZE_IMAGE
2020-01-13T16:59:46.5793750Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:256:8: note: 'CALIB_CB_NORMALIZE_IMAGE' declared here
2020-01-13T16:59:46.5793850Z        CALIB_CB_NORMALIZE_IMAGE = 2,
2020-01-13T16:59:46.5793920Z        ^
2020-01-13T16:59:46.6458220Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:628:45: error: use of undeclared identifier 'CV_CALIB_CB_ADAPTIVE_THRESH'; did you mean 'CALIB_CB_ADAPTIVE_THRESH'?
2020-01-13T16:59:46.6458860Z                                             CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.6458990Z                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.6459090Z                                             CALIB_CB_ADAPTIVE_THRESH
2020-01-13T16:59:46.6459650Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:255:8: note: 'CALIB_CB_ADAPTIVE_THRESH' declared here
2020-01-13T16:59:46.6459760Z enum { CALIB_CB_ADAPTIVE_THRESH = 1,
2020-01-13T16:59:46.6459830Z        ^
2020-01-13T16:59:46.6460400Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:628:75: error: use of undeclared identifier 'CV_CALIB_CB_FAST_CHECK'; did you mean 'CALIB_CB_FAST_CHECK'?
2020-01-13T16:59:46.6460580Z                                             CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.6460680Z                                                                           ^~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.6460780Z                                                                           CALIB_CB_FAST_CHECK
2020-01-13T16:59:46.6461260Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:258:8: note: 'CALIB_CB_FAST_CHECK' declared here
2020-01-13T16:59:46.6461370Z        CALIB_CB_FAST_CHECK      = 8,
2020-01-13T16:59:46.6461450Z        ^
2020-01-13T16:59:46.6462020Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:628:100: error: use of undeclared identifier 'CV_CALIB_CB_NORMALIZE_IMAGE'; did you mean 'CALIB_CB_NORMALIZE_IMAGE'?
2020-01-13T16:59:46.6462370Z                                             CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.6462460Z                                                                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.6462570Z                                                                                                    CALIB_CB_NORMALIZE_IMAGE
2020-01-13T16:59:46.6463100Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:256:8: note: 'CALIB_CB_NORMALIZE_IMAGE' declared here
2020-01-13T16:59:46.6463200Z        CALIB_CB_NORMALIZE_IMAGE = 2,
2020-01-13T16:59:46.6463280Z        ^
2020-01-13T16:59:46.6619530Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:643:17: error: use of undeclared identifier 'CV_CALIB_FIX_ASPECT_RATIO'; did you mean 'CALIB_FIX_ASPECT_RATIO'?
2020-01-13T16:59:46.6619780Z     if( flags & CV_CALIB_FIX_ASPECT_RATIO )
2020-01-13T16:59:46.6619900Z                 ^~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.6619980Z                 CALIB_FIX_ASPECT_RATIO
2020-01-13T16:59:46.6620490Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:270:8: note: 'CALIB_FIX_ASPECT_RATIO' declared here
2020-01-13T16:59:46.6620590Z        CALIB_FIX_ASPECT_RATIO    = 0x00002,
2020-01-13T16:59:46.6620660Z        ^
2020-01-13T16:59:46.6803700Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:653:40: error: use of undeclared identifier 'CV_CALIB_FIX_K3'; did you mean 'CALIB_FIX_K3'?
2020-01-13T16:59:46.6803920Z                     Dist, rvecs, tvecs,CV_CALIB_FIX_K3);
2020-01-13T16:59:46.6804010Z                                        ^~~~~~~~~~~~~~~
2020-01-13T16:59:46.6804080Z                                        CALIB_FIX_K3
2020-01-13T16:59:46.6804580Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:276:8: note: 'CALIB_FIX_K3' declared here
2020-01-13T16:59:46.6804710Z        CALIB_FIX_K3              = 0x00080,
2020-01-13T16:59:46.6804790Z        ^
2020-01-13T16:59:46.7122340Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:714:51: error: use of undeclared identifier 'CV_CALIB_CB_ADAPTIVE_THRESH'; did you mean 'CALIB_CB_ADAPTIVE_THRESH'?
2020-01-13T16:59:46.7122770Z                                                   CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.7122880Z                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.7122960Z                                                   CALIB_CB_ADAPTIVE_THRESH
2020-01-13T16:59:46.7123720Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:255:8: note: 'CALIB_CB_ADAPTIVE_THRESH' declared here
2020-01-13T16:59:46.7123840Z enum { CALIB_CB_ADAPTIVE_THRESH = 1,
2020-01-13T16:59:46.7123930Z        ^
2020-01-13T16:59:46.7124630Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:714:81: error: use of undeclared identifier 'CV_CALIB_CB_NORMALIZE_IMAGE'; did you mean 'CALIB_CB_NORMALIZE_IMAGE'?
2020-01-13T16:59:46.7124750Z                                                   CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_NORMALIZE_IMAGE);
2020-01-13T16:59:46.7124850Z                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.7124950Z                                                                                 CALIB_CB_NORMALIZE_IMAGE
2020-01-13T16:59:46.7125480Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:256:8: note: 'CALIB_CB_NORMALIZE_IMAGE' declared here
2020-01-13T16:59:46.7125580Z        CALIB_CB_NORMALIZE_IMAGE = 2,
2020-01-13T16:59:46.7125650Z        ^
2020-01-13T16:59:46.7730350Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:767:25: error: use of undeclared identifier 'CV_CALIB_FIX_ASPECT_RATIO'; did you mean 'CALIB_FIX_ASPECT_RATIO'?
2020-01-13T16:59:46.7730740Z                         CV_CALIB_FIX_ASPECT_RATIO + CV_CALIB_ZERO_TANGENT_DIST + CV_CALIB_SAME_FOCAL_LENGTH + CV_CALIB_FIX_K3,
2020-01-13T16:59:46.7730830Z                         ^~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.7730910Z                         CALIB_FIX_ASPECT_RATIO
2020-01-13T16:59:46.7731440Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:270:8: note: 'CALIB_FIX_ASPECT_RATIO' declared here
2020-01-13T16:59:46.7731540Z        CALIB_FIX_ASPECT_RATIO    = 0x00002,
2020-01-13T16:59:46.7731620Z        ^
2020-01-13T16:59:46.7732170Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:767:53: error: use of undeclared identifier 'CV_CALIB_ZERO_TANGENT_DIST'; did you mean 'CALIB_ZERO_TANGENT_DIST'?
2020-01-13T16:59:46.7732300Z                         CV_CALIB_FIX_ASPECT_RATIO + CV_CALIB_ZERO_TANGENT_DIST + CV_CALIB_SAME_FOCAL_LENGTH + CV_CALIB_FIX_K3,
2020-01-13T16:59:46.7732380Z                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.7732470Z                                                     CALIB_ZERO_TANGENT_DIST
2020-01-13T16:59:46.7732930Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:272:8: note: 'CALIB_ZERO_TANGENT_DIST' declared here
2020-01-13T16:59:46.7733030Z        CALIB_ZERO_TANGENT_DIST   = 0x00008,
2020-01-13T16:59:46.7733100Z        ^
2020-01-13T16:59:46.7733800Z /Users/runner/runners/2.163.1/work/robotology-superbuild/robotology-superbuild/robotology/ICUB/src/tools/stereoCalib/src/stereoCalibThread.cpp:767:82: error: use of undeclared identifier 'CV_CALIB_SAME_FOCAL_LENGTH'; did you mean 'CALIB_SAME_FOCAL_LENGTH'?
2020-01-13T16:59:46.7733960Z                         CV_CALIB_FIX_ASPECT_RATIO + CV_CALIB_ZERO_TANGENT_DIST + CV_CALIB_SAME_FOCAL_LENGTH + CV_CALIB_FIX_K3,
2020-01-13T16:59:46.7734060Z                                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~
2020-01-13T16:59:46.7734520Z                                                                                  CALIB_SAME_FOCAL_LENGTH
2020-01-13T16:59:46.7735070Z /usr/local/Cellar/opencv/4.2.0_1/include/opencv4/opencv2/calib3d.hpp:289:8: note: 'CALIB_SAME_FOCAL_LENGTH' declared here
2020-01-13T16:59:46.7735160Z        CALIB_SAME_FOCAL_LENGTH   = 0x00200,
2020-01-13T16:59:46.7735240Z        ^
2020-01-13T16:59:46.7735640Z fatal error: too many errors emitted, stopping now [-ferror-limit=]
2020-01-13T16:59:46.7852270Z 20 errors generated.
2020-01-13T16:59:46.7948790Z make[5]: *** [src/tools/stereoCalib/CMakeFiles/stereoCalib.dir/src/stereoCalibThread.cpp.o] Error 1
2020-01-13T16:59:46.7956150Z make[4]: *** [src/tools/stereoCalib/CMakeFiles/stereoCalib.dir/all] Error 2
2020-01-13T16:59:46.7963400Z make[3]: *** [all] Error 2
2020-01-13T16:59:46.7978370Z make[2]: *** [robotology/ICUB/CMakeFiles/YCMStamp/ICUB-build] Error 2
2020-01-13T16:59:46.7981270Z make[1]: *** [CMakeFiles/ICUB.dir/all] Error 2
2020-01-13T16:59:46.7989190Z make: *** [all] Error 2
2020-01-13T16:59:46.8018810Z ##[error]Process completed with exit code 2.
2020-01-13T16:59:46.8107470Z Cleaning up orphan processes
~~~

I noticed also that compiling `icub-main` with opencv 4 there is a linking problem

```
CMakeFiles/stereoCalib.dir/src/main.cpp.o: In function `cvflann::anyimpl::big_any_policy<cv::String>::static_delete(void**)':
main.cpp:(.text._ZN7cvflann7anyimpl14big_any_policyIN2cv6StringEE13static_deleteEPPv[_ZN7cvflann7anyimpl14big_any_policyIN2cv6StringEE13static_deleteEPPv]+0x15): undefined reference to `cv::String::deallocate()'
CMakeFiles/stereoCalib.dir/src/main.cpp.o: In function `cvflann::anyimpl::big_any_policy<cv::String>::move(void* const*, void**)':
main.cpp:(.text._ZN7cvflann7anyimpl14big_any_policyIN2cv6StringEE4moveEPKPvPS5_[_ZN7cvflann7anyimpl14big_any_policyIN2cv6StringEE4moveEPKPvPS5_]+0x10): undefined reference to `cv::String::deallocate()'
main.cpp:(.text._ZN7cvflann7anyimpl14big_any_policyIN2cv6StringEE4moveEPKPvPS5_[_ZN7cvflann7anyimpl14big_any_policyIN2cv6StringEE4moveEPKPvPS5_]+0x24): undefined reference to `cv::String::deallocate()'
CMakeFiles/stereoCalib.dir/src/stereoCalibThread.cpp.o: In function `cv::FileStorage& cv::operator<< <cv::Mat>(cv::FileStorage&, cv::Mat const&) [clone .constprop.201]':
stereoCalibThread.cpp:(.text+0x179): undefined reference to `cv::write(cv::FileStorage&, cv::String const&, cv::Mat const&)'
stereoCalibThread.cpp:(.text+0x1cd): undefined reference to `cv::String::allocate(unsigned long)'
stereoCalibThread.cpp:(.text+0x227): undefined reference to `cv::error(int, cv::String const&, char const*, char const*, int)'
stereoCalibThread.cpp:(.text+0x22f): undefined reference to `cv::String::deallocate()'
stereoCalibThread.cpp:(.text+0x244): undefined reference to `cv::String::deallocate()'

``` 

This is due to the fact that the content of `${OpenCV_LIBRARIES}` changed in opencv 4.

## OpenCV 3
```
opencv_calib3d
opencv_core
opencv_features2d
opencv_flann
opencv_highgui
opencv_imgcodecs
opencv_imgproc
opencv_ml
opencv_objdetect
opencv_photo
opencv_shape
opencv_stitching
opencv_superres
opencv_video
opencv_videoio
opencv_videostab
opencv_viz
opencv_aruco
opencv_bgsegm
opencv_bioinspired
opencv_ccalib
opencv_datasets
opencv_dpm
opencv_face
opencv_freetype
opencv_fuzzy
opencv_hdf
opencv_line_descriptor
opencv_optflow
opencv_phase_unwrapping
opencv_plot
opencv_reg
opencv_rgbd
opencv_saliency
opencv_stereo
opencv_structured_light
opencv_surface_matching
opencv_text
opencv_ximgproc
opencv_xobjdetect
opencv_xphoto
opencv_calib3d
opencv_core
opencv_features2d
opencv_flann
opencv_highgui
opencv_imgcodecs
opencv_imgproc
opencv_ml
opencv_objdetect
opencv_photo
opencv_shape
opencv_stitching
opencv_superres
opencv_video
opencv_videoio
opencv_videostab
opencv_viz
opencv_aruco
opencv_bgsegm
opencv_bioinspired
opencv_ccalib
opencv_datasets
opencv_dpm
opencv_face
opencv_freetype
opencv_fuzzy
opencv_hdf
opencv_line_descriptor
opencv_optflow
opencv_phase_unwrapping
opencv_plot
opencv_reg
opencv_rgbd
opencv_saliency
opencv_stereo
opencv_structured_light
opencv_surface_matching
opencv_text
opencv_ximgproc
opencv_xobjdetect
opencv_xphoto
```

## OpenCV 4
```
opencv_video
opencv_imgproc
opencv_features2d
opencv_videoio
opencv_stitching
opencv_imgcodecs
opencv_calib3d
opencv_core
opencv_photo
opencv_highgui
opencv_dnn
opencv_gapi
opencv_ml
opencv_flann
opencv_objdetect
opencv_bioinspired
opencv_ximgproc
opencv_hfs
opencv_stereo
opencv_dpm
opencv_phase_unwrapping
opencv_videostab
opencv_face
opencv_tracking
opencv_hdf
opencv_img_hash
opencv_shape
opencv_superres
opencv_text
opencv_xfeatures2d
opencv_freetype
opencv_structured_light
opencv_reg
opencv_datasets
opencv_surface_matching
opencv_sfm
opencv_dnn_objdetect
opencv_saliency
opencv_xobjdetect
opencv_bgsegm
opencv_xphoto
opencv_optflow
opencv_cvv
opencv_line_descriptor
opencv_ccalib
opencv_aruco
opencv_fuzzy
opencv_rgbd
opencv_plot

```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -44,7 +44,7 @@ jobs:
         git clone https://github.com/microsoft/vcpkg.git --depth 1 --branch 2019.10 ${VCPKG_ROBOTOLOGY_ROOT}
         git clone https://github.com/robotology/robotology-vcpkg-binary-ports.git ${VCPKG_ROBOTOLOGY_BIN_PORTS_ROOT}
         ${VCPKG_ROBOTOLOGY_ROOT}/bootstrap-vcpkg.sh
-        ${VCPKG_ROBOTOLOGY_ROOT}/vcpkg.exe --overlay-ports=${VCPKG_ROBOTOLOGY_BIN_PORTS_ROOT} install --triplet x64-windows ace freeglut ode sdl1 gsl eigen3 libxml2 sdl1 opencv3 ipopt-binary
+        ${VCPKG_ROBOTOLOGY_ROOT}/vcpkg.exe --overlay-ports=${VCPKG_ROBOTOLOGY_BIN_PORTS_ROOT} install --triplet x64-windows ace freeglut ode sdl1 gsl eigen3 libxml2 sdl1 opencv ipopt-binary
         rm -rf ${VCPKG_ROBOTOLOGY_ROOT}/buildtrees ${VCPKG_ROBOTOLOGY_ROOT}/packages ${VCPKG_ROBOTOLOGY_ROOT}/downloads
         ${VCPKG_ROBOTOLOGY_ROOT}/vcpkg.exe --overlay-ports=${VCPKG_ROBOTOLOGY_BIN_PORTS_ROOT} install --triplet x64-windows qt5-base
         rm -rf ${VCPKG_ROBOTOLOGY_ROOT}/buildtrees ${VCPKG_ROBOTOLOGY_ROOT}/packages ${VCPKG_ROBOTOLOGY_ROOT}/downloads
@@ -52,7 +52,7 @@ jobs:
     - name: Dependencies [macOS]
       if: matrix.os == 'macOS-latest'
       run: |
-        brew install ace cmake eigen gsl ipopt opencv@3 pkg-config qt5
+        brew install ace cmake eigen gsl ipopt opencv pkg-config qt5
     
     - name: Dependencies [Ubuntu]
       if: matrix.os == 'ubuntu-latest'

--- a/src/modules/actionsRenderingEngine/include/iCub/VisuoThread.h
+++ b/src/modules/actionsRenderingEngine/include/iCub/VisuoThread.h
@@ -28,7 +28,7 @@
 #include <yarp/sig/Vector.h>
 #include <yarp/sig/Image.h>
 
-#include <cv.h>
+#include <opencv2/core/core_c.h>
 
 #include <mutex>
 #include <string>

--- a/src/modules/camCalib/include/iCub/CamCalibModule.h
+++ b/src/modules/camCalib/include/iCub/CamCalibModule.h
@@ -13,7 +13,7 @@
 #include <stdio.h>
 
 // opencv
-#include <cv.h>
+#include <opencv2/core/core_c.h>
 
 // yarp
 #include <yarp/os/all.h>

--- a/src/modules/camCalib/include/iCub/PinholeCalibTool.h
+++ b/src/modules/camCalib/include/iCub/PinholeCalibTool.h
@@ -16,7 +16,9 @@
 #include <math.h>
 
 // opencv
-#include <cv.h>
+#include <opencv2/core/core_c.h>
+#include <opencv2/calib3d/calib3d_c.h>
+#include <opencv2/calib3d.hpp>
 
 // yarp
 //#include <yarp/sig/Image.h>

--- a/src/modules/camCalib/include/iCub/SphericalCalibTool.h
+++ b/src/modules/camCalib/include/iCub/SphericalCalibTool.h
@@ -16,7 +16,7 @@
 #include <yarp/os/Value.h>
 
 // opencv
-#include <cv.h>
+#include <opencv2/core/core_c.h>
 
 // iCub
 #include <iCub/ICalibTool.h>

--- a/src/modules/camCalib/src/PinholeCalibTool.cpp
+++ b/src/modules/camCalib/src/PinholeCalibTool.cpp
@@ -179,8 +179,9 @@ bool PinholeCalibTool::init(CvSize currImgSize, CvSize calibImgSize){
     _mapUndistortY = cvCreateImage(currImgSize, IPL_DEPTH_32F, 1);
 
     /* init the undistortion matrices */
-    cvInitUndistortMap( _intrinsic_matrix_scaled, _distortion_coeffs,
-                        _mapUndistortX, _mapUndistortY);
+    cv::initUndistortRectifyMap(cv::cvarrToMat(_intrinsic_matrix_scaled), cv::cvarrToMat(_distortion_coeffs), cv::Mat(),
+                                cv::cvarrToMat(_intrinsic_matrix_scaled), cv::Size(currImgSize.width, currImgSize.height),
+                                CV_32FC1,cv::cvarrToMat(_mapUndistortX), cv::cvarrToMat(_mapUndistortY));
 
     _needInit = false;
     return true;

--- a/src/modules/camCalib/src/main.cpp
+++ b/src/modules/camCalib/src/main.cpp
@@ -133,7 +133,7 @@
 #include <iCub/CamCalibModule.h>
 
 // OpenCV
-#include <cv.h>
+#include <opencv2/core/core_c.h>
 
 using namespace std;
 using namespace yarp::os;

--- a/src/modules/camCalibWithPose/include/iCub/CamCalibModule.h
+++ b/src/modules/camCalibWithPose/include/iCub/CamCalibModule.h
@@ -15,7 +15,7 @@
 #include <map>
 
 // opencv
-#include <cv.h>
+#include <opencv2/core/core_c.h>
 
 // yarp
 #include <yarp/os/all.h>

--- a/src/modules/camCalibWithPose/include/iCub/PinholeCalibTool.h
+++ b/src/modules/camCalibWithPose/include/iCub/PinholeCalibTool.h
@@ -16,7 +16,9 @@
 #include <math.h>
 
 // opencv
-#include <cv.h>
+#include <opencv2/core/core_c.h>
+#include <opencv2/calib3d/calib3d_c.h>
+#include <opencv2/calib3d.hpp>
 
 // yarp
 //#include <yarp/sig/Image.h>

--- a/src/modules/camCalibWithPose/include/iCub/SphericalCalibTool.h
+++ b/src/modules/camCalibWithPose/include/iCub/SphericalCalibTool.h
@@ -16,7 +16,7 @@
 #include <yarp/os/Value.h>
 
 // opencv
-#include <cv.h>
+#include <opencv2/core/core_c.h>
 
 // iCub
 #include <iCub/ICalibTool.h>

--- a/src/modules/camCalibWithPose/src/PinholeCalibTool.cpp
+++ b/src/modules/camCalibWithPose/src/PinholeCalibTool.cpp
@@ -179,8 +179,9 @@ bool PinholeCalibTool::init(CvSize currImgSize, CvSize calibImgSize){
     _mapUndistortY = cvCreateImage(currImgSize, IPL_DEPTH_32F, 1);
 
     /* init the undistortion matrices */
-    cvInitUndistortMap( _intrinsic_matrix_scaled, _distortion_coeffs,
-                        _mapUndistortX, _mapUndistortY);
+    cv::initUndistortRectifyMap(cv::cvarrToMat(_intrinsic_matrix_scaled), cv::cvarrToMat(_distortion_coeffs), cv::Mat(),
+                                cv::cvarrToMat(_intrinsic_matrix_scaled), cv::Size(currImgSize.width, currImgSize.height),
+                                CV_32FC1,cv::cvarrToMat(_mapUndistortX), cv::cvarrToMat(_mapUndistortY));
 
     _needInit = false;
     return true;

--- a/src/modules/camCalibWithPose/src/main.cpp
+++ b/src/modules/camCalibWithPose/src/main.cpp
@@ -133,7 +133,7 @@
 #include <iCub/CamCalibModule.h>
 
 // OpenCV
-#include <cv.h>
+#include <opencv2/core/core_c.h>
 
 using namespace std;
 using namespace yarp::os;

--- a/src/modules/dualCamCalib/include/iCub/DualCamCalibModule.h
+++ b/src/modules/dualCamCalib/include/iCub/DualCamCalibModule.h
@@ -8,7 +8,7 @@
 #define __DUALCAMCALIBMODULE__
 
 // opencv
-#include <cv.h>
+#include <opencv2/core/core_c.h>
 
 // yarp
 #include <yarp/os/BufferedPort.h>

--- a/src/modules/dualCamCalib/include/iCub/PinholeCalibTool.h
+++ b/src/modules/dualCamCalib/include/iCub/PinholeCalibTool.h
@@ -16,7 +16,9 @@
 #include <math.h>
 
 // opencv
-#include <cv.h>
+#include <opencv2/core/core_c.h>
+#include <opencv2/calib3d/calib3d_c.h>
+#include <opencv2/calib3d.hpp>
 
 // yarp
 //#include <yarp/sig/Image.h>

--- a/src/modules/dualCamCalib/include/iCub/SphericalCalibTool.h
+++ b/src/modules/dualCamCalib/include/iCub/SphericalCalibTool.h
@@ -16,7 +16,7 @@
 #include <yarp/os/Value.h>
 
 // opencv
-#include <cv.h>
+#include <opencv2/core/core_c.h>
 
 // iCub
 #include <iCub/ICalibTool.h>

--- a/src/modules/dualCamCalib/src/PinholeCalibTool.cpp
+++ b/src/modules/dualCamCalib/src/PinholeCalibTool.cpp
@@ -179,9 +179,9 @@ bool PinholeCalibTool::init(CvSize currImgSize, CvSize calibImgSize){
     _mapUndistortY = cvCreateImage(currImgSize, IPL_DEPTH_32F, 1);
 
     /* init the undistortion matrices */
-    cvInitUndistortMap( _intrinsic_matrix_scaled, _distortion_coeffs,
-                        _mapUndistortX, _mapUndistortY);
-
+    cv::initUndistortRectifyMap(cv::cvarrToMat(_intrinsic_matrix_scaled), cv::cvarrToMat(_distortion_coeffs), cv::Mat(),
+                                cv::cvarrToMat(_intrinsic_matrix_scaled), cv::Size(currImgSize.width, currImgSize.height),
+                                CV_32FC1,cv::cvarrToMat(_mapUndistortX), cv::cvarrToMat(_mapUndistortY));
     _needInit = false;
     return true;
 }

--- a/src/modules/dualCamCalib/src/main.cpp
+++ b/src/modules/dualCamCalib/src/main.cpp
@@ -16,7 +16,7 @@
 #include <iCub/DualCamCalibModule.h>
 
 // OpenCV
-#include <cv.h>
+#include <opencv2/core/core_c.h>
 
 using namespace std;
 using namespace yarp::os;

--- a/src/modules/templatePFTracker/include/iCub/particleFilter.h
+++ b/src/modules/templatePFTracker/include/iCub/particleFilter.h
@@ -40,9 +40,8 @@
 #include <iostream>
 #include <iomanip>
 #include <deque>
-
-#include <cv.h>
-#include <highgui.h>
+#include <opencv2/core/core_c.h>
+#include <opencv2/imgproc/imgproc_c.h>
 
 /* default number of particles */
 #define PARTICLES 1000

--- a/src/tools/depth2kin/src/module.cpp
+++ b/src/tools/depth2kin/src/module.cpp
@@ -26,6 +26,7 @@
 #include <utility>
 
 #include <yarp/cv/Cv.h>
+#include <opencv2/imgproc/imgproc_c.h>
 #include <yarp/math/Math.h>
 #include <yarp/math/Rand.h>
 #include <yarp/math/NormRand.h>

--- a/src/tools/stereoCalib/include/stereoCalibThread.h
+++ b/src/tools/stereoCalib/include/stereoCalibThread.h
@@ -3,7 +3,13 @@
 #include <string>
 #include <mutex>
 
-#include <opencv2/opencv.hpp>
+#include <opencv2/calib3d/calib3d_c.h>
+#include <opencv2/calib3d.hpp>
+#include <opencv2/core/core_c.h>
+#include <opencv2/core/types_c.h>
+#include <opencv2/imgcodecs.hpp>
+
+
 
 #include <yarp/os/all.h>
 #include <yarp/dev/all.h>

--- a/src/tools/stereoCalib/src/stereoCalibThread.cpp
+++ b/src/tools/stereoCalib/src/stereoCalibThread.cpp
@@ -830,7 +830,7 @@ void stereoCalibThread::saveCalibration(const string& extrinsicFilePath, const s
             return;
     }
 
-    FileStorage fs(intrinsicFilePath+".yml", CV_STORAGE_WRITE);
+    FileStorage fs(intrinsicFilePath+".yml", cv::FileStorage::Mode::WRITE);
     if( fs.isOpened() )
     {
         fs << "M1" << Kleft << "D1" << DistL << "M2" << Kright << "D2" << DistR;
@@ -839,7 +839,7 @@ void stereoCalibThread::saveCalibration(const string& extrinsicFilePath, const s
     else
         cout << "Error: can not save the intrinsic parameters\n";
 
-    fs.open(extrinsicFilePath+".yml", CV_STORAGE_WRITE);
+    fs.open(extrinsicFilePath+".yml", cv::FileStorage::Mode::WRITE);
     if( fs.isOpened() )
     {
         fs << "R" << R << "T" << T <<"Q" << Q;

